### PR TITLE
Add checkpoint coordination utility

### PR DESCRIPTION
**Pull request by HerbertJordan**
This PR introduces a new utility to the backend library facilitating the coordinated creation of checkpoints.

The utility provides an implementation of a `CheckpointCoordinator` capable of synchronizing the creation of a checkpoint for a range of generic `CheckpointParticipant`s. To facilitate consistent creation of checkpoints, the coordinator is implementing the [Two-phase commit protocol](https://en.wikipedia.org/wiki/Two-phase_commit_protocol).

This is the first step towards resolving #946. A full prototype implementation is provided by #970. However, #970 is too large to be reviewed in a single PR. This is the first self-contained part of it.

### CODE REVIEW

**Review by kjezek — DISMISSED** [review]
I put some comments and questions next to the code, but otherwise I think it looks good.

File: go/backend/utils/checkpoint.go [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+}
+
+// CheckpointParticipant is a participant in a checkpoint that can participate in
+// the coordinated creation and restoration of checkpoints.
+type CheckpointParticipant interface {
+	// GuaranteeCheckpoint requires the participant to check whether a restoration
+	// to the given checkpoint is possible. If the participant is not able to restore to
+	// the given checkpoint, an error is returned. If the participant finds the given
+	// checkpoint as a prepared checkpoint that has not yet been committed, it should
+	// be committed. If there is a prepared checkpoint beyond the given checkpoint,
```
[the remark this answers is not in the document]
Comment (HerbertJordan) [reply go/backend/utils/checkpoint.go:56]: Updated the documentation. 

The intention is that this function should do a cleanup after a restart to get to the given checkpoint. If the current checkpoint is already the given checkpoint, everything is fine. If the given checkpoint is a pending checkpoint, this pending checkpoint should be committed. If there is a pending checkpoint that is not the given checkpoint, this checkpoint can be discarded.

File: go/backend/utils/checkpoint.go [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+}
+
+func readCheckpointFile(path string) (Checkpoint, error) {
+	file, err := os.Open(path)
+	if err != nil {
+		return 0, err
+	}
+	content := make([]byte, 4)
+	if _, err := io.ReadFull(file, content); err != nil {
+		return 0, err
```
[the remark this answers is not in the document]
Comment (HerbertJordan) [reply go/backend/utils/checkpoint.go:214]: Thanks, good find. The file is closed now.

I would not want to use `ReadFile` without checking how large the file is.

File: go/backend/utils/checkpoint.go [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+		prepared = append(prepared, p)
+	}
+
+	// Prepare and commit the checkpoint file atomically.
+	prepareFile := filepath.Join(c.path, "prepare")
+	commitFile := filepath.Join(c.path, "committed")
+	err := errors.Join(
+		createCheckpointFile(prepareFile, commit),
+		// Renaming files in the same directory is (in most cases) atomic on POSIX systems.
+		os.Rename(prepareFile, commitFile),
```
[the remark this answers is not in the document]
Comment (HerbertJordan) [reply go/backend/utils/checkpoint.go:167]: At this point all participants have prepared their commits to the extend that they are able to perform the commit or revert to the previous commit. This atomic rename operation is the decision point. If the system crashes before this operation, everything is reverted back to the previous checkpoint the next time the data structures are opened. If the system crashes after, everything is moving to the pending checkpoints on recovery.

File: go/backend/utils/checkpoint.go [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+//
+// Use of this software is governed by the Business Source License included
+// in the LICENSE file and at fantom.foundation/bsl11.
+//
+// Change Date: 2028-4-16
+//
+// On the date above, in accordance with the Business Source License, use of
+// this software will be governed by the GNU Lesser General Public License v3.
+
+package utils
```
[the remark this answers is not in the document]
Comment (HerbertJordan) [reply go/backend/utils/checkpoint.go]: Done

**Review by kjezek — APPROVED** [review]
looks good, thanks

**Pull request merged**